be/src/exec/hash_join_components.cpp

```diff
@@ -85,6 +98,228 @@ void HashJoinProber::attach(HashJoinBuilder* builder, const HashJoinProbeMetrics
+        if (_partition_input_channels[i].is_empty()) {
+            continue;
+        }
+        if (!probers[i]->probe_chunk_empty()) {
+            continue;
+        }
+        RETURN_IF_ERROR(probers[i]->push_probe_chunk(runtime_state, _partition_input_channels[i].pull()));
+    }
+    return Status::OK();
+}
+
+Status PartitionedHashJoinProberImpl::push_probe_chunk(RuntimeState* state, ChunkPtr&& chunk) {
+    SCOPED_TIMER(_hash_joiner.probe_metrics().partition_probe_overhead);
+    auto& probers = _probers;
+    auto& partition_keys = _hash_joiner.probe_expr_ctxs();
+
+    size_t num_rows = chunk->num_rows();
+    size_t num_partitions = probers.size();
+    size_t num_partition_cols = partition_keys.size();
+
+    std::vector<ColumnPtr> partition_columns(num_partition_cols);
+    for (size_t i = 0; i < num_partition_cols; ++i) {
+        ASSIGN_OR_RETURN(partition_columns[i], partition_keys[i]->evaluate(chunk.get()));
+    }
+    std::vector<uint32_t> hash_values;
+    {
+        hash_values.assign(num_rows, HashUtil::FNV_SEED);
+
+        for (const ColumnPtr& column : partition_columns) {
+            column->fnv_hash(hash_values.data(), 0, num_rows);
+        }
+        // find partition id
+        for (size_t i = 0; i < hash_values.size(); ++i) {
+            hash_values[i] = HashUtil::fmix32(hash_values[i]) & (num_partitions - 1);
+        }
+    }
+
+    const auto& partitions = hash_values;
+
+    std::vector<uint32_t> selection;
+    selection.resize(chunk->num_rows());
+
+    std::vector<int32_t> channel_row_idx_start_points;
+    channel_row_idx_start_points.assign(num_partitions + 1, 0);
+
+    for (uint32_t i : partitions) {
+        channel_row_idx_start_points[i]++;
+    }
+
+    for (int32_t i = 1; i <= channel_row_idx_start_points.size() - 1; ++i) {
+        channel_row_idx_start_points[i] += channel_row_idx_start_points[i - 1];
+    }
+
+    for (int32_t i = chunk->num_rows() - 1; i >= 0; --i) {
+        selection[channel_row_idx_start_points[partitions[i]] - 1] = i;
+        channel_row_idx_start_points[partitions[i]]--;
+    }
+    _partition_input_channels.resize(num_partitions);
+
+    for (size_t i = 0; i < num_partitions; ++i) {
+        auto from = channel_row_idx_start_points[i];
+        auto size = channel_row_idx_start_points[i + 1] - from;
+        if (size == 0) {
+            continue;
+        }
+
+        if (_partition_input_channels[i].is_empty()) {
+            _partition_input_channels[i].push(chunk->clone_empty());
+        }
+
+        if (_partition_input_channels[i].back()->num_rows() + size <= 4096) {
+            _partition_input_channels[i].back()->append_selective(*chunk, selection.data(), from, size);
+        } else {
+            _partition_input_channels[i].push(chunk->clone_empty());
+            _partition_input_channels[i].back()->append_selective(*chunk, selection.data(), from, size);
+        }
+
+        if (_partition_input_channels[i].is_full()) {
```

Comment:
if there is quite large binary column, the number of partitions are large,  each probe channel is almost full, the HashJoin would use up num_partitions fold of non-optimized version, so is_full should take bytes into consideration to avoid OOM.

Reply:
another question is that partitioned chunk in probe channel can also occupy L2 cache, should we consider that 
L2 occupied by probe channel undergoing processing + bytes of corresponding partitioned HT < byte size L2 cache ?

Reply:
> if there is quite large binary column, the number of partitions are large, each probe channel is almost full, the HashJoin would use up num_partitions fold of non-optimized version, so is_full should take bytes into consideration to avoid OOM.

I'll add an extra check for mem limit



> another question is that partitioned chunk in probe channel can also occupy L2 cache, should we consider that L2 occupied by probe channel undergoing processing + bytes of corresponding partitioned HT < byte size L2 cache ?
 
According to the test “Case outputs the right table and the left table at the same time” I don't find this obvious. I think it's that the probe chunk is always accessed sequentially, so the effect on cache is limited.